File: app/src/main/java/com/kollectivemobile/euki/manager/ContentManagerImpl.java
package com.kollectivemobile.euki.manager;

import android.util.Log;

import com.kollectivemobile.euki.model.ContentItem;
import com.kollectivemobile.euki.networking.EukiCallback;
import com.kollectivemobile.euki.utils.Utils;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class ContentManagerImpl implements ContentManager {
    private static final String ABORTION_KNOWLEDGE_JSON = "abortion_knowledge";
    private static final String ABORTION_WALKTHROUGH_JSON = "abortion_walkthrough";
    private static final String STIS_JSON = "stis";
    private static final String CONTRACEPTION_JSON = "contraception";
    private static final String MISCARRIAGE_JSON = "miscarriage";
    private static final String SEX_RELATIONSHIPS_JSON = "sex_relationships";
    private static final String PREGNANCY_OPTIONS = "pregnancy_options";
    private static final String MENSTRUATION_OPTIONS = "menstruation_options";

    private ContentItem mAbortionKnowledge;
    private ContentItem mAbortionWalkthrough;
    private ContentItem mSTI;
    private ContentItem mContraception;
    private ContentItem mMiscarriage;
    private ContentItem mSexRelationships;
    private ContentItem mPregnancyOptions;
    private ContentItem mMenstruationOptions;

    private AbortionContentManager mAbortionContentManager;
    private PrivacyContentManager mPrivacyContentManager;

    public ContentManagerImpl(AbortionContentManager abortionContentManager, PrivacyContentManager privacyContentManager) {
        this.mAbortionContentManager = abortionContentManager;
        this.mPrivacyContentManager = privacyContentManager;
    }

    @Override
    public void getAbortionKnowledge(EukiCallback<ContentItem> callback) {
        callback.onSuccess(getAbortionKnowledge());
    }

    @Override
    public void getAbortionWalkthrough(EukiCallback<ContentItem> callback) {
        callback.onSuccess(getAbortionWalkthrough());
    }

    @Override
    public void getSTIs(EukiCallback<ContentItem> callback) {
        callback.onSuccess(getSTI());
    }

    @Override
    public void getContraception(EukiCallback<ContentItem> callback) {
        callback.onSuccess(getContraception());
    }

    @Override
    public void getMiscarriage(EukiCallback<ContentItem> callback) {
        callback.onSuccess(getMiscarriage());
    }

    @Override
    public void getSexRelationships(EukiCallback<ContentItem> callback) {
        callback.onSuccess(getSexRelationships());
    }

    @Override
    public void getPregnancyOptions(EukiCallback<ContentItem> callback) {
        callback.onSuccess(getPregnancyOptions());
    }

    @Override
    public void getMenstruationOptions(EukiCallback<ContentItem> callback) {
        callback.onSuccess(getMenstruationOptions());
    }

    @Override
    public ContentItem getContentItem(String id) {
        ContentItem foundContentItem = getContentItem(id, getAbortionKnowledge());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, getAbortionWalkthrough());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, getSTI());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, getContraception());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, getMiscarriage());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, getSexRelationships());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, getPregnancyOptions());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, getMenstruationOptions());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        // Abortion ContentItems

        foundContentItem = getContentItem(id, mAbortionContentManager.getAbortionMifeMiso12());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, mAbortionContentManager.getAbortionMiso12());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, mAbortionContentManager.getAbortionSuctionVacuum());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, mAbortionContentManager.getAbortionDilationEvacuation());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, mAbortionContentManager.getMedialAbortion());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        // Privacy ContentItems

        foundContentItem = getContentItem(id, mPrivacyContentManager.getPrivacyFAQs());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        foundContentItem = getContentItem(id, mPrivacyContentManager.getPrivacyStatement());
        if (foundContentItem != null) {
            return foundContentItem;
        }

        return null;
    }

    @Override
    public ContentItem getContentItem(String id, ContentItem contentItem) {
        if (contentItem == null || contentItem.getId() == null) {
            return null;
        }

        if (contentItem.getId().equals(id)) {
            return contentItem;
        }

        if (contentItem.getContentItems() != null) {
            for (ContentItem childItem : contentItem.getContentItems()) {
                ContentItem foundItem = getContentItem(id, childItem);
                if (foundItem != null) {
                    return foundItem;
                }
            }
        }

        if (contentItem.getExpandableItems() != null) {
            for (ContentItem childItem : contentItem.getExpandableItems()) {
                ContentItem foundItem = getContentItem(id, childItem);
                if (foundItem != null) {
                    foundItem.setParent(contentItem);
                    return foundItem;
                }
            }
        }

        if (contentItem.getSelectableItems() != null) {
            for (ContentItem childItem : contentItem.getSelectableItems()) {
                ContentItem foundItem = getContentItem(id, childItem);
                if (foundItem != null) {
                    return foundItem;
                }
            }
        }

        if (contentItem.getSelectableRowItems() != null) {
            for (ContentItem childItem : contentItem.getSelectableRowItems()) {
                ContentItem foundItem = getContentItem(id, childItem);
                if (foundItem != null) {
                    return foundItem;
                }
            }
        }

        return null;
    }

    @Override
    public List<ContentItem> searchContentItem(String searchText) {
        List<ContentItem> items = new ArrayList<>();

        items.addAll(searchContentItem(searchText, getAbortionKnowledge()));
        items.addAll(searchContentItem(searchText, getAbortionWalkthrough()));
        items.addAll(searchContentItem(searchText, getSTI()));
        items.addAll(searchContentItem(searchText, getContraception()));
        items.addAll(searchContentItem(searchText, getMiscarriage()));
        items.addAll(searchContentItem(searchText, getSexRelationships()));
        items.addAll(searchContentItem(searchText, getPregnancyOptions()));
        items.addAll(searchContentItem(searchText, getMenstruationOptions()));

        //Abortion ContentItems

        items.addAll(searchContentItem(searchText, mAbortionContentManager.getAbortionMifeMiso12()));
        items.addAll(searchContentItem(searchText, mAbortionContentManager.getAbortionMiso12()));
        items.addAll(searchContentItem(searchText, mAbortionContentManager.getAbortionSuctionVacuum()));
        items.addAll(searchContentItem(searchText, mAbortionContentManager.getAbortionDilationEvacuation()));
        items.addAll(searchContentItem(searchText, mAbortionContentManager.getMedialAbortion()));

        List<ContentItem> foundItems = new ArrayList<>();

        for (ContentItem item : items) {
            if (!foundItems.contains(item)) {
                foundItems.add(item);
            }
        }

        return foundItems;
    }

    @Override
    public List<ContentItem> searchContentItem(String searchText, ContentItem contentItem) {
        List<ContentItem> items = new ArrayList<>();

        if (contentItem.getId() == null) {
            return items;
        }

        String id = Utils.getLocalized(contentItem.getId());
        if (id != null && id.toLowerCase().contains(searchText.toLowerCase())) {
            items.add(contentItem);
        }

        String title = Utils.getLocalized(contentItem.getTitle());
        if (title != null && title.toLowerCase().contains(searchText.toLowerCase())) {
            items.add(contentItem);
        }

        String content = Utils.getLocalized(contentItem.getContent());
        if (content != null && content.toLowerCase().contains(searchText.toLowerCase())) {
            items.add(contentItem);
        }

        if (contentItem.getContentItems() != null) {
            for (ContentItem childItem : contentItem.getContentItems()) {
                if (searchContentItem(searchText, childItem).size() > 0) {
                    items.add(contentItem);
                }
            }
        }

        if (contentItem.getExpandableItems() != null) {
            for (ContentItem childItem : contentItem.getExpandableItems()) {
                if (searchContentItem(searchText, childItem).size() > 0) {
                    items.add(contentItem);
                }
            }
        }

        if (contentItem.getSelectableItems() != null) {
            for (ContentItem childItem : contentItem.getSelectableItems()) {
                List<ContentItem> foundItems = searchContentItem(searchText, childItem);
                items.addAll(foundItems);
            }
        }

        if (contentItem.getSelectableRowItems() != null) {
            for (ContentItem childItem : contentItem.getSelectableRowItems()) {
                List<ContentItem> foundItems = searchContentItem(searchText, childItem);
                items.addAll(foundItems);
            }
        }

        List<ContentItem> foundItems = new ArrayList<>();

        for (ContentItem item : items) {
            if (!foundItems.contains(item)) {
                foundItems.add(item);
            }
        }

        return foundItems;
    }

    private ContentItem createContentItem(String filename) {
        JSONObject jsonObject = Utils.getJsonFromAssets(filename);
        return new ContentItem(jsonObject);
    }

    /*
        Gettters
     */

    public ContentItem getAbortionKnowledge() {
        if (mAbortionKnowledge == null) {
            mAbortionKnowledge = createContentItem(ABORTION_KNOWLEDGE_JSON);
        }
        return mAbortionKnowledge;
    }

    public ContentItem getAbortionWalkthrough() {
        if (mAbortionWalkthrough == null) {
            mAbortionWalkthrough = createContentItem(ABORTION_WALKTHROUGH_JSON);
        }
        return mAbortionWalkthrough;
    }

    public ContentItem getSTI() {
        if (mSTI == null) {
            mSTI = createContentItem(STIS_JSON);
        }
        return mSTI;
    }

    public ContentItem getContraception() {
        if (mContraception == null) {
            mContraception = createContentItem(CONTRACEPTION_JSON);
        }
        return mContraception;
    }

    public ContentItem getMiscarriage() {
        if (mMiscarriage == null) {
            mMiscarriage = createContentItem(MISCARRIAGE_JSON);
        }
        return mMiscarriage;
    }

    public ContentItem getSexRelationships() {
        if (mSexRelationships == null) {
            mSexRelationships = createContentItem(SEX_RELATIONSHIPS_JSON);
        }
        return mSexRelationships;
    }

    public ContentItem getPregnancyOptions() {
        if (mPregnancyOptions == null) {
            mPregnancyOptions = createContentItem(PREGNANCY_OPTIONS);
        }
        return mPregnancyOptions;
    }

    public ContentItem getMenstruationOptions() {
        if (mMenstruationOptions == null) {
            mMenstruationOptions = createContentItem(MENSTRUATION_OPTIONS);
        }
        return mMenstruationOptions;
    }
}

File: app/src/main/java/com/kollectivemobile/euki/manager/ContentManagerImpl.kt
package com.kollectivemobile.euki.manager

import com.kollectivemobile.euki.model.ContentItem
import com.kollectivemobile.euki.networking.EukiCallback
import com.kollectivemobile.euki.utils.Utils

class ContentManagerImpl(
        private val mAbortionContentManager: AbortionContentManager,
        private val mPrivacyContentManager: PrivacyContentManager
) : ContentManager {

  companion object {
    private const val ABORTION_KNOWLEDGE_JSON = "abortion_knowledge"
    private const val ABORTION_WALKTHROUGH_JSON = "abortion_walkthrough"
    private const val STIS_JSON = "stis"
    private const val CONTRACEPTION_JSON = "contraception"
    private const val MISCARRIAGE_JSON = "miscarriage"
    private const val SEX_RELATIONSHIPS_JSON = "sex_relationships"
    private const val PREGNANCY_OPTIONS = "pregnancy_options"
    private const val MENSTRUATION_OPTIONS = "menstruation_options"
  }

  private var mAbortionKnowledge: ContentItem? = null
  private var mAbortionWalkthrough: ContentItem? = null
  private var mSTI: ContentItem? = null
  private var mContraception: ContentItem? = null
  private var mMiscarriage: ContentItem? = null
  private var mSexRelationships: ContentItem? = null
  private var mPregnancyOptions: ContentItem? = null
  private var mMenstruationOptions: ContentItem? = null

  override fun getAbortionKnowledge(callback: EukiCallback<ContentItem>) {
    callback.onSuccess(getAbortionKnowledge())
  }

  override fun getAbortionWalkthrough(callback: EukiCallback<ContentItem>) {
    callback.onSuccess(getAbortionWalkthrough())
  }

  override fun getSTIs(callback: EukiCallback<ContentItem>) {
    callback.onSuccess(getSTI())
  }

  override fun getContraception(callback: EukiCallback<ContentItem>) {
    callback.onSuccess(getContraception())
  }

  override fun getMiscarriage(callback: EukiCallback<ContentItem>) {
    callback.onSuccess(getMiscarriage())
  }

  override fun getSexRelationships(callback: EukiCallback<ContentItem>) {
    callback.onSuccess(getSexRelationships())
  }

  override fun getPregnancyOptions(callback: EukiCallback<ContentItem>) {
    callback.onSuccess(getPregnancyOptions())
  }

  override fun getMenstruationOptions(callback: EukiCallback<ContentItem>) {
    callback.onSuccess(getMenstruationOptions())
  }

  override fun getContentItem(id: String): ContentItem? {
    // Search through main content items
    val mainContentItems =
            listOf(
                    getAbortionKnowledge(),
                    getAbortionWalkthrough(),
                    getSTI(),
                    getContraception(),
                    getMiscarriage(),
                    getSexRelationships(),
                    getPregnancyOptions(),
                    getMenstruationOptions()
            )

    for (contentItem in mainContentItems) {
      val foundContentItem = getContentItem(id, contentItem)
      if (foundContentItem != null) {
        return foundContentItem
      }
    }

    // Search through abortion content items
    val abortionContentItems =
            listOf(
                    mAbortionContentManager.abortionMifeMiso12,
                    mAbortionContentManager.abortionMiso12,
                    mAbortionContentManager.abortionSuctionVacuum,
                    mAbortionContentManager.abortionDilationEvacuation,
                    mAbortionContentManager.medialAbortion
            )

    for (contentItem in abortionContentItems) {
      val foundContentItem = getContentItem(id, contentItem)
      if (foundContentItem != null) {
        return foundContentItem
      }
    }

    // Search through privacy content items
    val privacyContentItems =
            listOf(
                    mPrivacyContentManager.privacyFAQs,
                    mPrivacyContentManager.privacyStatement,
                    mPrivacyContentManager.privacyBestPractices
            )

    for (contentItem in privacyContentItems) {
      val foundContentItem = getContentItem(id, contentItem)
      if (foundContentItem != null) {
        return foundContentItem
      }
    }

    return null
  }

  override fun getContentItem(id: String, contentItem: ContentItem): ContentItem? {
    if (contentItem.id == null) {
      return null
    }

    if (contentItem.id == id) {
      return contentItem
    }

    // Search in content items
    contentItem.contentItems?.forEach { childItem ->
      val foundItem = getContentItem(id, childItem)
      if (foundItem != null) {
        return foundItem
      }
    }

    // Search in expandable items
    contentItem.expandableItems?.forEach { childItem ->
      val foundItem = getContentItem(id, childItem)
      if (foundItem != null) {
        foundItem.parent = contentItem
        return foundItem
      }
    }

    // Search in selectable items
    contentItem.selectableItems?.forEach { childItem ->
      val foundItem = getContentItem(id, childItem)
      if (foundItem != null) {
        return foundItem
      }
    }

    // Search in selectable row items
    contentItem.selectableRowItems?.forEach { childItem ->
      val foundItem = getContentItem(id, childItem)
      if (foundItem != null) {
        return foundItem
      }
    }

    return null
  }

  override fun searchContentItem(searchText: String): List<ContentItem> {
    val items = mutableListOf<ContentItem>()

    // Search main content items
    val mainContentItems =
            listOf(
                    getAbortionKnowledge(),
                    getAbortionWalkthrough(),
                    getSTI(),
                    getContraception(),
                    getMiscarriage(),
                    getSexRelationships(),
                    getPregnancyOptions(),
                    getMenstruationOptions()
            )

    for (contentItem in mainContentItems) {
      items.addAll(searchContentItem(searchText, contentItem))
    }

    // Search abortion content items
    val abortionContentItems =
            listOf(
                    mAbortionContentManager.abortionMifeMiso12,
                    mAbortionContentManager.abortionMiso12,
                    mAbortionContentManager.abortionSuctionVacuum,
                    mAbortionContentManager.abortionDilationEvacuation,
                    mAbortionContentManager.medialAbortion
            )

    for (contentItem in abortionContentItems) {
      items.addAll(searchContentItem(searchText, contentItem))
    }

    // Remove duplicates
    return items.distinct()
  }

  override fun searchContentItem(searchText: String, contentItem: ContentItem): List<ContentItem> {
    val items = mutableListOf<ContentItem>()

    if (contentItem.id == null) {
      return items
    }

    val searchTextLower = searchText.lowercase()

    // Check if current item matches search criteria
    var matches = false

    Utils.getLocalized(contentItem.id)?.let { id ->
      if (id.lowercase().contains(searchTextLower)) {
        matches = true
      }
    }

    Utils.getLocalized(contentItem.title)?.let { title ->
      if (title.lowercase().contains(searchTextLower)) {
        matches = true
      }
    }

    Utils.getLocalized(contentItem.content)?.let { content ->
      if (content.lowercase().contains(searchTextLower)) {
        matches = true
      }
    }

    if (matches) {
      items.add(contentItem)
    }

    // Search in child items
    contentItem.contentItems?.forEach { childItem ->
      if (searchContentItem(searchText, childItem).isNotEmpty()) {
        items.add(contentItem)
      }
    }

    contentItem.expandableItems?.forEach { childItem ->
      if (searchContentItem(searchText, childItem).isNotEmpty()) {
        items.add(contentItem)
      }
    }

    contentItem.selectableItems?.forEach { childItem ->
      items.addAll(searchContentItem(searchText, childItem))
    }

    contentItem.selectableRowItems?.forEach { childItem ->
      items.addAll(searchContentItem(searchText, childItem))
    }

    // Remove duplicates
    return items.distinct()
  }

  private fun createContentItem(filename: String): ContentItem {
    val jsonObject = Utils.getJsonFromAssets(filename)
    return ContentItem(jsonObject)
  }

  /*
     Getters
  */

  private fun getAbortionKnowledge(): ContentItem {
    if (mAbortionKnowledge == null) {
      mAbortionKnowledge = createContentItem(ABORTION_KNOWLEDGE_JSON)
    }
    return mAbortionKnowledge!!
  }

  private fun getAbortionWalkthrough(): ContentItem {
    if (mAbortionWalkthrough == null) {
      mAbortionWalkthrough = createContentItem(ABORTION_WALKTHROUGH_JSON)
    }
    return mAbortionWalkthrough!!
  }

  private fun getSTI(): ContentItem {
    if (mSTI == null) {
      mSTI = createContentItem(STIS_JSON)
    }
    return mSTI!!
  }

  private fun getContraception(): ContentItem {
    if (mContraception == null) {
      mContraception = createContentItem(CONTRACEPTION_JSON)
    }
    return mContraception!!
  }

  private fun getMiscarriage(): ContentItem {
    if (mMiscarriage == null) {
      mMiscarriage = createContentItem(MISCARRIAGE_JSON)
    }
    return mMiscarriage!!
  }

  private fun getSexRelationships(): ContentItem {
    if (mSexRelationships == null) {
      mSexRelationships = createContentItem(SEX_RELATIONSHIPS_JSON)
    }
    return mSexRelationships!!
  }

  private fun getPregnancyOptions(): ContentItem {
    if (mPregnancyOptions == null) {
      mPregnancyOptions = createContentItem(PREGNANCY_OPTIONS)
    }
    return mPregnancyOptions!!
  }

  private fun getMenstruationOptions(): ContentItem {
    if (mMenstruationOptions == null) {
      mMenstruationOptions = createContentItem(MENSTRUATION_OPTIONS)
    }
    return mMenstruationOptions!!
  }
}
